site/src/components/Modal/LoginWithLoginCodeScreen.tsx

```diff
@@ -0,0 +1,325 @@
+import { SerializedUser } from "../../lib/model/user.model";
+import {
+  ApiLoginWithLoginCodeRequestBody,
+  ApiLoginWithLoginCodeResponse,
+} from "../../pages/api/loginWithLoginCode.api";
+import { Button } from "../Button";
+import { TextField } from "../TextField";
+import {
+  isVerificationCodeFormatted,
+  useVerificationCodeTextField,
+} from "../hooks/useVerificationCodeTextField";
+
+type LoginWithLoginCodeScreenProps = {
+  userId: string;
+  loginCodeId: string;
+  email: string;
+  initialVerificationCode?: string;
+  initialApiLoginErrorMessage?: ReactNode;
+  setLoginCodeId: (loginCodeId: string) => void;
+  onLogin: (user: SerializedUser) => void;
+  onChangeEmail: () => void;
+};
+
+const parseVerificationCodeInput = (inputCode: string) =>
+  inputCode.replace(/\s/g, "");
+
+export const LoginWithLoginCodeScreen: VFC<LoginWithLoginCodeScreenProps> = ({
+  userId,
+  loginCodeId,
+  email,
+  initialVerificationCode,
+  initialApiLoginErrorMessage,
+  setLoginCodeId,
+  onLogin,
+  onChangeEmail,
+}) => {
+  const loginCodeInputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (loginCodeInputRef.current) {
+      loginCodeInputRef.current.select();
+    }
+  }, [loginCodeInputRef]);
+
+  const [touchedVerificationCodeInput, setTouchedVerificationCodeInput] =
+    useState<boolean>(false);
+
+  const [sendingLoginCode, setSendingLoginCode] = useState<boolean>(false);
+  const [loggingIn, setLoggingIn] = useState<boolean>(false);
+
+  const [apiResendEmailErrorMessage, setApiResendEmailErrorMessage] =
+    useState<ReactNode>();
+  const [apiLoginErrorMessage, setApiLoginErrorMessage] = useState<ReactNode>(
+    initialApiLoginErrorMessage,
+  );
+
+  const {
+    verificationCode,
+    setVerificationCode,
+    isVerificationCodeValid: isVerificationCodeInputValid,
+    verificationCodeHelperText,
+  } = useVerificationCodeTextField({ initialVerificationCode });
+
+  useEffect(() => {
+    if (initialApiLoginErrorMessage) {
+      setApiLoginErrorMessage(initialApiLoginErrorMessage);
+    }
+  }, [initialApiLoginErrorMessage]);
+
+  useEffect(() => {
+    if (initialVerificationCode) {
+      setTouchedVerificationCodeInput(true);
+      setVerificationCode(initialVerificationCode);
+    }
+  }, [initialVerificationCode, setVerificationCode]);
+
+  const handleResendEmail = async () => {
+    setSendingLoginCode(true);
+    setApiResendEmailErrorMessage(undefined);
+
+    const { data, error } = await apiClient.sendLoginCode({ email });
+    setSendingLoginCode(false);
+
+    if (error) {
+      if (error.response?.data.errors) {
+        setApiResendEmailErrorMessage(
+          error.response.data.errors.map(({ msg }) => msg),
+        );
+      } else {
+        throw error;
+      }
+    } else if (data) {
+      setVerificationCode("");
+      setApiLoginErrorMessage(undefined);
+      setTouchedVerificationCodeInput(false);
+      setLoginCodeId(data.loginCodeId);
+      if (loginCodeInputRef.current) {
```

Comment:
This will lead to multiple re-renders. React (<v18) batches updates in event handlers but that doesn't apply when the handler is an async function.

> However, React wasn’t consistent about when it batches updates. For example, if you need to fetch data, and then update the state in the handleClick above, then React would not batch the updates, and perform two independent updates.
This is because React used to only batch updates during a browser event (like click), but here we’re updating the state after the event has already been handled (in fetch callback)

[Link](https://github.com/reactwg/react-18/discussions/21#discussion-3385721)

It might be better if we handle state management with useReducer. Can also use `unstable_batchedUpdates` to fix the immediate issue if we don't want to go through `useReducer` route

Reply:
Good spot, thank you. I'm going to opt for `unstable_batchedUpdates` for now, as I don't want to invest too much time into perfecting how the state is stored here as it will likely change once we come up with a common approach for capturing the `TextField` related state.

Reply:
But if I'm understanding correctly this will be resolved once we move to React > 18 anyway

Reply:
Addressed in: https://github.com/blockprotocol/blockprotocol/pull/84/commits/4d73656ae7318724825a036574c4e486f6e5d9e7